Fix cpp extension compile failure on some envs

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37221 Fix cpp extension compile failure on some envs**

CRTP template pattern sometimes can make nvcc compiler confused when calling class member's member function without explicitly adding 'this->'. This caused cloneable.h not compatible with some env setup. 

you will get compile error like the following:
```
torch/nn/cloneable.h:44:65: error: invalid static_cast from type ‘const torch::OrderedDict<std::basic_string<char>, at::Tensor>’ to type ‘torch::OrderedDict<std::basic_string<char>, at::Tensor>&’
torch/nn/cloneable.h:56:59: error: invalid static_cast from type ‘const torch::OrderedDict<std::basic_string<char>, at::Tensor>’ to type ‘torch::OrderedDict<std::basic_string<char>, at::Tensor>&’
torch/nn/cloneable.h:68:61: error: invalid static_cast from type ‘const torch::OrderedDict<std::basic_string<char>, std::shared_ptr<torch::nn::Module> >’ to type ‘torch::OrderedDict<std::basic_string<char>, std::shared_ptr<torch::nn::Module> >&’
```

The affected env is mainly for ubuntu16.04: 
eg.
ubuntu 16.04 + cuda 9.2 + gcc 5 (tested with 5.4 here)
ubuntu 16.04 + cuda 9.2 + gcc 7 (tested with 7.5 here)
ubuntu 16.04 + cuda 10.0 + gcc 5 (tested with 5.4 here)

This is a fix for such problem.

Differential Revision: [D21226873](https://our.internmc.facebook.com/intern/diff/D21226873)

Commit:
[Test only]Fix cpp extension compile failure on some envs

[ghstack-poisoned]

Commit:
Update on "[Test only]Fix cpp extension compile failure on some envs"


CRTP template pattern sometimes can get compiler confused when calling class member's member function without 'adding this->'. This caused cloneable.h not compatible with some env setup. eg. ubuntu 16.04 + cuda 9.2 + gcc 5.4.

Differential Revision: [D21226873](https://our.internmc.facebook.com/intern/diff/D21226873)

[ghstack-poisoned]

Commit:
Update on "Fix cpp extension compile failure on some envs"


CRTP template pattern sometimes can get compiler confused when calling class member's member function without 'adding this->'. This caused cloneable.h not compatible with some env setup. eg. ubuntu 16.04 + cuda 9.2 + gcc 5.4.

Differential Revision: [D21226873](https://our.internmc.facebook.com/intern/diff/D21226873)

[ghstack-poisoned]

gchanan commented on torch/csrc/api/include/torch/nn/cloneable.h:
```diff
@@ -42,7 +42,7 @@ class Cloneable : public virtual Module {
     copy->reset();
     TORCH_CHECK(
-        copy->parameters_.size() == parameters_.size(),
+        copy->parameters_.size() == this->parameters_.size(),
```
comment on all these places why you did it or people might accidentally remove it.

Commit:
Update on "Fix cpp extension compile failure on some envs"


CRTP template pattern sometimes can make nvcc compiler confused when calling class member's member function without explicitly adding 'this->'. This caused cloneable.h not compatible with some env setup. 

you will get compile error like the following:
```
torch/nn/cloneable.h:44:65: error: invalid static_cast from type ‘const torch::OrderedDict<std::basic_string<char>, at::Tensor>’ to type ‘torch::OrderedDict<std::basic_string<char>, at::Tensor>&’
torch/nn/cloneable.h:56:59: error: invalid static_cast from type ‘const torch::OrderedDict<std::basic_string<char>, at::Tensor>’ to type ‘torch::OrderedDict<std::basic_string<char>, at::Tensor>&’
torch/nn/cloneable.h:68:61: error: invalid static_cast from type ‘const torch::OrderedDict<std::basic_string<char>, std::shared_ptr<torch::nn::Module> >’ to type ‘torch::OrderedDict<std::basic_string<char>, std::shared_ptr<torch::nn::Module> >&’
```

The affected env is mainly for ubuntu16.04: 
eg.
ubuntu 16.04 + cuda 9.2 + gcc 5 (tested with 5.4 here)
ubuntu 16.04 + cuda 9.2 + gcc 7 (tested with 7.5 here)
ubuntu 16.04 + cuda 10.0 + gcc 5 (tested with 5.4 here)

This is a fix for such problem.

Differential Revision: [D21226873](https://our.internmc.facebook.com/intern/diff/D21226873)

[ghstack-poisoned]

Commit:
Update on "Fix cpp extension compile failure on some envs"


CRTP template pattern sometimes can make nvcc compiler confused when calling class member's member function without explicitly adding 'this->'. This caused cloneable.h not compatible with some env setup. 

you will get compile error like the following:
```
torch/nn/cloneable.h:44:65: error: invalid static_cast from type ‘const torch::OrderedDict<std::basic_string<char>, at::Tensor>’ to type ‘torch::OrderedDict<std::basic_string<char>, at::Tensor>&’
torch/nn/cloneable.h:56:59: error: invalid static_cast from type ‘const torch::OrderedDict<std::basic_string<char>, at::Tensor>’ to type ‘torch::OrderedDict<std::basic_string<char>, at::Tensor>&’
torch/nn/cloneable.h:68:61: error: invalid static_cast from type ‘const torch::OrderedDict<std::basic_string<char>, std::shared_ptr<torch::nn::Module> >’ to type ‘torch::OrderedDict<std::basic_string<char>, std::shared_ptr<torch::nn::Module> >&’
```

The affected env is mainly for ubuntu16.04: 
eg.
ubuntu 16.04 + cuda 9.2 + gcc 5 (tested with 5.4 here)
ubuntu 16.04 + cuda 9.2 + gcc 7 (tested with 7.5 here)
ubuntu 16.04 + cuda 10.0 + gcc 5 (tested with 5.4 here)

This is a fix for such problem.

Differential Revision: [D21226873](https://our.internmc.facebook.com/intern/diff/D21226873)

[ghstack-poisoned]
